crates/red_knot_python_semantic/src/types.rs

```diff
@@ -6061,22 +6059,59 @@ impl<'db> FunctionType<'db> {
         let scope = self.body_scope(db);
         let function_stmt_node = scope.node(db).expect_function();
         let definition = self.definition(db);
-        Signature::from_function(db, self.generic_context(db), definition, function_stmt_node)
+        let generic_context = function_stmt_node.type_params.as_ref().map(|type_params| {
+            let index = semantic_index(db, scope.file(db));
+            GenericContext::from_type_params(db, index, type_params)
+        });
+        let mut signature = Signature::from_function(
+            db,
+            generic_context,
+            self.inherited_generic_context(db),
+            definition,
+            function_stmt_node,
+        );
+        if let Some(specialization) = self.specialization(db) {
+            signature = signature.apply_specialization(db, specialization);
+        }
+        signature
     }
 
     pub(crate) fn is_known(self, db: &'db dyn Db, known_function: KnownFunction) -> bool {
         self.known(db) == Some(known_function)
     }
 
-    fn with_generic_context(self, db: &'db dyn Db, generic_context: GenericContext<'db>) -> Self {
+    fn with_dataclass_transformer_params(
+        self,
+        db: &'db dyn Db,
+        params: DataclassTransformerParams,
+    ) -> Self {
+        Self::new(
```

Comment:
Reasonable to debug-assert here that it's `None`, or is this not an invariant in this case?

Reply:
I don't think it's an invariant, since then a user could trigger a panic by applying `dataclass_transform` twice.  That could be a diagnostic we want to enforce, but shouldn't be an assertion.  (Inheriting a generic context at most once is a proper invariant, since it's not possible for a user to declare a function "doubly" generic)